
<div style="border:solid green 2px; padding: 20px">

👋 **¡Hola! Soy Dot, tu revisor de IA.**

He completado la primera revisión de tu código. A continuación, encontrarás mis comentarios y sugerencias de mejora.

**¿Qué debes hacer ahora?**

1. **Revisar:** Lee mis comentarios en el notebook más abajo.
2. **Decidir:**
* **¿Estás de acuerdo con el feedback?** ¡Genial! Actualiza tu código según las sugerencias.
* **¿No estás de acuerdo o crees que es un error?** ¡No hay problema! Puedes dejar tu código exactamente como está.


3. **Volver a enviar:** Envía tu proyecto de nuevo. **Tu próxima iteración será revisada por un experto humano**, independientemente de si hiciste cambios o no.

-------------------------------------------------------------------------------------------------------------
Mis comentarios están más abajo. **Te pido amablemente que no los muevas, modifiques ni elimines**.

Verás mis comentarios resaltados en cuadros verdes, amarillos o rojos como estos:

<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration X</b> <a class="tocSkip"></a>

Correcto. Todo se ha hecho exitosamente.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration X</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment, Iteration X</b> <a class="tocSkip"></a>

Necesita corrección. El bloque requiere algunas correcciones.
</div>



<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

# Comentario General 2

Hola Daniel, espero que te encuentres muy bien.  

Hiciste un proyecto excelente!

Abordaste todos los detalles y solucionaste correctamente todo.

Ademas, tus observaciones y conclusiones fueron detalladas y fundamentadas.

Puedo dar por aprobado tu proyecto, un excelente proyecto.

Sigue así!

</div>


# Déjame escuchar música

Los datos están almacenados en el archivo `/datasets/music_project_en.csv`.

¡Ahora sí, manos al código!


## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [27]:
import pandas as pd

<details>
<summary>Haz clic para ver la pista</summary>

Para usar `pandas`, necesitas la palabra clave `import` seguida del nombre de la biblioteca y un alias.


Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [28]:
df =  pd.read_csv('/datasets/music_project_en.csv')


<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Buen trabajo importando pandas con un alias estándar y cargando `/datasets/music_project_en.csv` en `df` con `read_csv()`. Esto te deja bien preparado/a para el resto del análisis.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa la palabra clave `read_csv` para leer el archivo CSV. Asegúrate de incluir la ruta `/datasets/music_project_en.csv` y guarda el resultado en la variable `df`.

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [29]:
print (df.head(10))


     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `head` para ver las primeras filas del DataFrame.

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [30]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `info` para obtener detalles sobre columnas, tipos de datos y valores nulos.


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver dos problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. `Los espacios en blanco.`




### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1.Los tipos de datos son `object`, significa que los tipos de datos están como cadenas de texto `strings` o una mezcla de tipos, con el uso de `df.head()` observamos que son canciones con datos como el artista, genero, hora y día en la que se reprodujo la canción. 

2.La cantidad de datos parace adecuada para establecer algún tipo de hipótesis y comprobarla, hay algunos planteamientos que no podrán ser abordados como  ¿cuánto tiempo escucharon una canción determinada? 

3.Tenemos valores ausentes en artist, track y genre; la columna time debería ser del tipo datetime pero está como objecto, y los errores ya mencionados en los encabzados del data frame 

<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Buen informe de revisión de datos: interpretaste los tipos `object`, comentaste si el dataset es suficiente para responder a las preguntas y señalaste problemas clave de calidad (valores faltantes y una columna que estaría mejor representada como un tipo de tiempo/datetime).

</div>


## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [31]:
print (df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


<div class="alert alert-block alert-danger">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Este paso está pensado para mostrar los nombres de las columnas originales (sin limpiar) antes de cualquier preprocesamiento. Dado que la salida ya muestra nombres limpios, parece que se ejecutó algún renombrado/limpieza anteriormente o que el notebook se ejecutó fuera de orden; considera reiniciar y ejecutar las celdas de arriba hacia abajo para que la progresión requerida sea verificable.

</div>


<div class="alert alert-block alert-success">
<b>Comentario de Revisor Iter 2</b> <a class="tocSkip"></a>

Muy bien, 

Ahora podemos ver los encabezados de la tabla!

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el atributo `.columns` para ver todos los nombres de las columnas en el DataFrame.


Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case, es decir, añade un guion bajo ( _ ) entre las palabras en lugar de un espacio.



Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [32]:
new_col_name = []
for old_name in df.columns:
    new_col_name.append(old_name.lower())
df.columns = new_col_name

print (df.columns.tolist())

['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day']


<div class="alert alert-block alert-success">
<b>Comentario de Revisor Iter 2</b> <a class="tocSkip"></a>

Excelente, 

Buen trabajo usando el bucle de for para modificar los encabezados con y `lower()`.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa un `for` para recorrer `df.columns` y aplica `.lower()` a cada nombre. Luego, asigna la lista resultante de nuevo a `df.columns`.


Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [33]:
new_col_name = []
for old_name in df.columns:
    new_col_name.append(old_name.strip())
df.columns = new_col_name

print (df.columns.tolist())

['userid', 'track', 'artist', 'genre', 'city', 'time', 'day']


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Buen flujo de trabajo de preprocesamiento: inspeccionaste los headers, luego los limpiaste paso a paso (pasándolos a minúsculas y eliminando espacios) y verificaste el resultado después de cada transformación, lo que hace que tus cambios sean fáciles de seguir y depurar.<br>
</div>

<details>
<summary>Haz clic para ver la pista</summary>

Usa `.strip()` dentro del bucle para limpiar los nombres de columnas. Asegúrate de reasignar los nuevos nombres a `df.columns`.


Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [34]:
df.rename(columns = {'userid':'user_id'}, inplace = True)

print (df.columns.tolist())

['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']


<div class="alert alert-block alert-success">
<b>Comentario de Revisor Iter 2</b> <a class="tocSkip"></a>

¡Excelente!

Realizas correctamente la limpieza de de los encabezados y además le das `snake_format` correctamente, bien hecho!
   
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `.rename()` con el parámetro `columns` para cambiar `userid` a `user_id`. No olvides establecer `inplace=True` o reasignar a `df`.

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [35]:
print (df.columns.tolist())


['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.columns` para volver a ver los nombres actualizados de las columnas.

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [36]:
print (df.isnull().sum())
print()
print (df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

La tarea te pide que demuestres dos métodos para contar valores faltantes. Ahora mismo solo muestras uno (`isnull().sum()`). Considera añadir un segundo enfoque equivalente (por ejemplo, usando `isna()` u otro método válido de conteo) y mostrar ambos resultados.

</div>


<div class="alert alert-block alert-success">
<b>Comentario de Revisor Iter 2</b> <a class="tocSkip"></a>

Hiciste un excelente trabajo acá en realidad.   
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.isna()` para detectar los valores ausentes y combínalo con el método `.sum()` para contar cuántos hay por columna.


Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [37]:
columns_to_replace = ['track', 'artist', 'genre']

for col in columns_to_replace:
    df[col] = df[col].fillna('unknown')


<div class="alert alert-block alert-danger">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Al rellenar los valores faltantes usaste `'unknow'` en lugar del requerido `'unknown'`. Esto introduce un valor de categoría adicional no deseado que se propagará en análisis posteriores (p. ej., en `genre` y `artist`).

</div>


<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration 2</b> <a class="tocSkip"></a>

Buen flujo de trabajo: reemplazaste los NaNs en las columnas especificadas y luego verificaste, con un recuento actualizado de valores faltantes, que no quedaran valores faltantes.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Crea una lista con los nombres de las columnas y usa un `for` para recorrerla. Dentro del bucle, aplica `.fillna('unknown')` a cada una.

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [38]:
df.isnull().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

<details>
<summary>Haz clic para ver la pista</summary>

Usa `isna()` con `sum()` otra vez para confirmar que ya no quedan valores nulos en las columnas modificadas.

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [39]:
print (df.duplicated().sum())
print()
print (len(df[df.duplicated()]))

3826

3826


<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Requisito recurrente: este paso también pide dos métodos para calcular el número de duplicados explícitos, pero solo se muestra uno (`duplicated().sum()`). Por favor, añade una segunda forma de calcular/confirmar el mismo recuento.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa un método que detecte filas repetidas (`duplicated`) y luego otro para contarlas.

Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [40]:
df.drop_duplicates(inplace = True)

<details>
<summary>Haz clic para ver la pista</summary>

Busca un método que elimine las filas duplicadas directamente del DataFrame. No olvides actualizar la variable.

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [41]:
df.duplicated().sum()

0

<div class="alert alert-block alert-success">
<b>Comentario de Revisor Iter 2</b> <a class="tocSkip"></a>
    
Muy buen uso del `drop_duplicates` para eliminar valores duplicados

Excelente trabajo!
   
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Repite el mismo enfoque que usaste para contar duplicados antes. Deberías ver que el total ahora es cero.


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.


Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [42]:
sorted(df['genre'].unique())

['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

<details>
<summary>Haz clic para ver la pista</summary>

Usa `unique()` para obtener los géneros únicos. Si quieres verlos en orden alfabético, combina con `sorted()`.

Etapa 2.12. Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_values()`.



1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:

* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.



In [43]:
def replace_wrong_values(df,column,wrong_values,correct_value):
    for wrong_value in wrong_values:
        df[column]=df[column].replace(wrong_value,correct_value)
    return df


<div class="alert alert-block alert-success">
<b>Comentario de Revisor Iter 2</b> <a class="tocSkip"></a>
    
Excelente trabajo desarrollando la funcion de replace_wrong_values   
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Crea una función con cuatro parámetros. Dentro del bucle, llama a `.replace()` sobre la columna para cada valor incorrecto, cambiándolo por el correcto.


Etapa 2.13. Ahora, llama a la función pasando:

* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto



In [44]:
replace_wrong_values(df, 'genre',['hip', 'hop', 'hip-hop'],'hiphop')


,user_id,track,artist,genre,city,time,day
0,FFB692EC,Kamigata To Boots,The Mass Missile,rock,Shelbyville,20:28:33,Wednesday
1,55204538,Delayed Because of Accident,Andreas Rönnberg,rock,Springfield,14:07:09,Friday
2,20EC38,Funiculì funiculà,Mario Lanza,pop,Shelbyville,20:58:07,Wednesday
3,A3DD03C9,Dragons in the Sunset,Fire + Ice,folk,Shelbyville,08:37:09,Monday
4,E2DC1FAE,Soul People,Space Echo,dance,Springfield,08:34:34,Monday
...,...,...,...,...,...,...,...
65074,729CBB09,My Name,McLean,rnb,Springfield,13:32:28,Wednesday
65075,D08D4A55,Maybe One Day (feat. Black Spade),Blu & Exile,hiphop,Shelbyville,10:00:00,Monday
65076,C5E3A0D5,Jalopiina,unknown,industrial,Springfield,20:09:26,Friday
65077,321D0506,Freight Train,Chas McDevitt,rock,Springfield,21:43:59,Friday


<details>
<summary>Haz clic para ver la pista</summary>

Llama a la función que creaste y pasa los cuatro argumentos necesarios. Asegúrate de usar el nombre correcto de la columna y lista los valores que deben unificarse.

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [45]:
sorted(df['genre'].unique())


['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

<div class="alert alert-block alert-success">
<b>Comentario de Revisor Iter 2</b> <a class="tocSkip"></a>

Excelente desarrollo impecable.

Defines correctamente la función para reemplazar valores erróneos y lo ejecutas correctamente.

Además, al mostrar los datos ordenas la tabla, excelente!
   
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa de nuevo el método que muestra los valores únicos de una columna para verificar que todos los nombres se hayan unificado correctamente.

## Etapa 3. Análisis

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

Queremos analizar si hay diferencias en la cantidad de canciones reproducidas en Springfield y Shelbyville. Para ello, usaremos los datos de dos días de la semana: lunes y viernes.

Compararemos cuántas canciones se escucharon en cada ciudad durante esos días para identificar posibles patrones de comportamiento.

Sigue estos tres pasos para organizar tu análisis:

- Dividir: agrupa los datos por ciudad.

- Aplicar: cuenta cuántas canciones se reproducen en cada grupo.

- Combinar: presenta los resultados de forma que se puedan comparar fácilmente ambas ciudades.

Repite este proceso por separado para cada uno de los dos días.


Etapa 3.1.
Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna `'track'` como referencia.


In [46]:
print (df.groupby('city')['track'].count())


city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Agrupa los datos por ciudad con el método `groupby()` y usa el método `count()` para contar las canciones reproducidas en cada grupo.

Etapa 3.2. Redacta brevemente tus observaciones sobre los resultados.

¿Qué diferencias encontraste entre Springfield y Shelbyville? ¿A qué podrían deberse?

Lo obvio es los números presentados, pero hace falta saber, densidad poblacional, edades promeido de la población, acceso a dispositivos moviles, conexión a internet, tipos de subscripción.

<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration 2</b> <a class="tocSkip"></a>

Interpretación clara y completa: comparaste Springfield vs. Shelbyville tanto para el lunes como para el viernes y diste explicaciones razonables para la diferencia en la actividad.

</div>



Etapa 3.3.
Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes y viernes.




In [47]:
print (df.groupby('day')['track'].count().loc[['Monday','Friday']])

day
Monday    21354
Friday    21840
Name: track, dtype: int64


<div class="alert alert-block alert-danger">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Este paso solicita recuentos por día de la semana solo para lunes y viernes. Tu salida agrupa por `city` y `day` e incluye miércoles, por lo que no coincide con el resultado solicitado—considera filtrar a lunes/viernes y agrupar por `day` para este paso en particular.

</div>


<div class="alert alert-block alert-success">
<b>Comentario de Revisor Iter 2</b> <a class="tocSkip"></a>

Excelente trabajo, esto en realidad está excelente!

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Cambia el criterio de agrupación a la columna que representa el día de la semana y filtra solo para los días que te interesan.

Etapa 3.4. Describe brevemente qué observaste al comparar los lunes y viernes.

¿Hubo un día con más actividad? ¿Cambia algo si analizas cada ciudad por separado?

El viernes tiene mayor actividada al se superior en 486 mas reproducciones

<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Tu comparación de Monday vs Friday es demasiado general. Sería más claro indicar explícitamente qué día tiene mayor actividad (en general y/o por ciudad) y respaldarlo con los conteos que calculaste.

</div>


Etapa 3.5

Ahora vamos a combinar dos criterios: día y ciudad.

Crea una función llamada `number_tracks()` que reciba dos parámetros:

* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)

Dentro de la función:

1. Filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Cuenta cuántas veces aparece `'user_id'` en ese filtro.
4. Devuelve ese número como resultado.



In [48]:

# Declara la función number_tracks() con dos parámetros: day= y city=.
# Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
# Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
# Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
# Devuelve el número de valores de la columna 'user_id'


def number_tracks(day, city):
    track_list = df[(df['day'] == day) & (df['city'] == city)]
    track_list_count = track_list['user_id'].count()
    return track_list_count


<div class="alert alert-block alert-success">
<b>Comentario de Revisor Iter 2</b> <a class="tocSkip"></a>

Buen trabajo!

La función `number_tracks` la defines excelentemente para recibir los argumentos `day` y `city`.

Excelente trabajo!

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Define una función con dos parámetros. Dentro, aplica un filtrado secuencial usando `df[...]`, y al final cuenta los valores con `.count()`.

Etapa 3.6. Llama a `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

In [49]:
# El número de canciones reproducidas en Springfield el lunes
number_tracks('Monday','Springfield')


15740

In [50]:
# El número de canciones reproducidas en Shelbyville el lunes
number_tracks('Monday','Shelbyville')


5614

In [51]:
# El número de canciones reproducidas en Springfield el viernes
number_tracks('Friday', 'Springfield')

15945

In [52]:
# El número de canciones reproducidas en Shelbyville el viernes
number_tracks('Friday', 'Shelbyville')

5895

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Buen trabajo: `number_tracks()` sigue la lógica solicitada (filtrar por día, luego por ciudad, contar `user_id`, devolver el resultado), y la llamaste para las cuatro combinaciones de día/ciudad para permitir una comparación clara.<br>
</div>

<details>
<summary>Haz clic para ver la pista</summary>

Llama a tu función con combinaciones distintas de ciudad y día, como `'Monday', 'Springfield'`, y guarda o imprime los resultados para comparar.

# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre el análisis

Redacta un resumen breve y claro de los hallazgos obtenidos durante el proceso de análisis. Tu conclusión debe:

* Mencionar los principales patrones que observaste en los datos.
* Identificar cualquier problema encontrado y cómo lo solucionaste.
* Explicar cómo estas acciones ayudaron a mejorar la calidad del análisis.

Reflexiona también sobre la pregunta central:

> *¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*

Apoya tu respuesta con ejemplos concretos de los resultados obtenidos.


1. Principales patrones observados

A través del análisis de la actividad de los usuarios en Springfield y Shelbyville, se identificaron comportamientos distintos según la ubicación y el tiempo:

Springfield: Es la ciudad con mayor volumen de reproducción en general. Muestra una actividad constante entre el inicio y el fin de semana, con 15,740 canciones los lunes y 15,945 los viernes.

Shelbyville: Presenta un volumen de usuarios significativamente menor. Su actividad aumenta ligeramente los viernes (5,895) en comparación con los lunes (5,614).

Miércoles: Se observó que, para Shelbyville, el miércoles es el día de mayor actividad con 7,003 reproducciones, mientras que para Springfield es el día más bajo con 11,056.

2. Problemas encontrados y soluciones

Durante el preprocesamiento de los datos, se resolvieron inconsistencias críticas que habrían invalidado las conclusiones:

Valores Ausentes: Se detectaron valores nulos en track, artist y genre. Se solucionó sustituyéndolos por el término 'unknown' para evitar la pérdida de registros en el análisis.

Duplicados Explícitos: Se eliminaron filas idénticas para asegurar que cada reproducción se contara una sola vez.

Duplicados Implícitos: En la columna genre, se detectó que el género "hiphop" estaba registrado erróneamente como hip, hop y hip-hop. Se creó y aplicó la función replace_wrong_values() para unificarlos bajo un solo nombre estándar.

3. Impacto en la calidad del análisis

Estas acciones de limpieza permitieron que los resultados finales fueran fiables y precisos:

Consistencia: Al unificar los nombres de los géneros, el conteo de estilos musicales ahora refleja la realidad sin fragmentar la popularidad de un género por errores de escritura.

Integridad: La gestión de valores ausentes y el ajuste de los encabezados de las columnas a formato snake_case facilitaron un procesamiento de datos libre de errores técnicos.

¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?

Sí, los datos confirman esta variación. Los resultados obtenidos con la función number_tracks() demuestran que los hábitos de escucha no son uniformes:

En Springfield, el consumo de música es masivo y estable entre lunes y viernes.

En Shelbyville, existe una preferencia marcada por los miércoles, superando significativamente su actividad de los lunes y viernes.

Esta diferencia sugiere que factores como la densidad poblacional o las rutinas laborales de cada ciudad influyen directamente en la forma en que los usuarios consumen música en la plataforma.

<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Conclusiones finales sólidas: están bien estructuradas, respaldadas con cifras concretas y vinculan claramente los pasos de limpieza (missing values, duplicados explícitos/implícitos, limpieza de encabezados) con una mayor fiabilidad del análisis. Buen trabajo al responder directamente a la pregunta central sobre las diferencias por ciudad y día.

</div>


<div class="alert alert-block alert-success">
<b>Comentario de Revisor Iter 2</b> <a class="tocSkip"></a>

Cierras con broche de oro un excelente proyecto, Daniel.

Tus conclusiones son muy acertadas y las fundamentas con datos reales del bloque de análisis.
   
</div>
